pkg/infra/iac2/templates_compiler.go

```diff
@@ -188,78 +196,133 @@ func (tc TemplatesCompiler) resolveDependencies(resource core.Resource) string {
 }
 
 // resolveStructInput translates a value to a form suitable to inject into the typescript as an input to a function.
-func (tc TemplatesCompiler) resolveStructInput(childVal reflect.Value, useDoubleQuotedStrings bool) string {
+func (tc TemplatesCompiler) resolveStructInput(childVal reflect.Value, useDoubleQuotedStrings bool, iacTag string) (string, error) {
 	var zeroValue reflect.Value
 	if childVal == zeroValue {
-		return `null`
+		return `null`, nil
 	}
 	switch childVal.Kind() {
 	case reflect.Bool,
 		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
 		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
 		reflect.Float32, reflect.Float64:
-		return fmt.Sprintf("%v", childVal.Interface())
+		return fmt.Sprintf("%v", childVal.Interface()), nil
 	case reflect.String:
-		return quoteTsString(childVal.Interface().(string), useDoubleQuotedStrings)
+		return quoteTsString(childVal.Interface().(string), useDoubleQuotedStrings), nil
 	case reflect.Struct, reflect.Pointer:
 		if childVal.Kind() == reflect.Pointer && childVal.IsNil() {
-			return "null"
+			return "null", nil
 		}
 		if typedChild, ok := childVal.Interface().(core.Resource); ok {
-			return tc.getVarName(typedChild)
+			return tc.getVarName(typedChild), nil
 		} else if typedChild, ok := childVal.Interface().(core.IaCValue); ok {
-			return tc.handleIaCValue(typedChild)
+			output, err := tc.handleIaCValue(typedChild)
+			if err != nil {
+				return output, err
+			}
+			return output, nil
 		} else {
-			return ""
+			if iacTag == "document" {
```

Comment:
I think this should be the first check in the if-else chain, not near the last.

If my struct tags a field as `resource:"document"`, then I'd expect that field to get rendered as a document no matter what, even if it also happens to implement `core.Resource` or `core.IaCValue`.

Reply:
that doesnt really make a ton of sense to me because if it is a Resource or an IaCValue we would completely break the iac by not returning what they do today, since in our minds they arent a nested document and instead they are something we know how to handle.

I see your point on that we would expect it to get handled, but we may need to think about that here.

Also then we have to look at each child value to see if it has the tag, where as now we only require having a top level tag because we know were rendering everything below as a child resource

Reply:
That all makes sense, and given that we don't actually need `core.Resource`s-as-documents today I'm fine with it. But then maybe just add a check above, and error out in case somebody does do it? Otherwise, I could see this being a pretty lengthy and frustrating debugging exercise in the future, if we do ever do have a `core.Resource`-as-document for some reason.

Reply:
Fair, ill go ahead and add the check for our supported resources then